components/plan-build-ps1/bin/hab-plan-build.ps1

```diff
@@ -0,0 +1,815 @@
+        $t = ConvertFrom-StringData -StringData (gc -raw .\results\last_build.env)
+
+        get-content -Encoding Byte .\results\fnichol-names-0.11.0-20161128001352-x86_64-windows.hart |
+            Select-Object -Skip 5 |
+            Set-Content -Encoding Byte -Path .\results\test.tar.xz
+        7z x -txz .\results\test.tar.xz .\results\test.tar
+        7z x -ttar .\results\test.tar .\results\hab
+    #>
+}
+
+# **Internal** Copy the final package artifact to the `$pkg_output_path`
+# directory as well as prepare a `last_build.env` report.
+function _Copy-BuildOutputs {
+    New-Item "$pkg_output_path" -ItemType Directory -Force | Out-Null
+    Copy-Item "$pkg_artifact" "$pkg_output_path"
+
+    $_pkg_sha256sum = (Get-FileHash "$pkg_artifact" -Algorithm SHA256).Hash.ToLower()
+    $_pkg_blake2bsum = $(& "$_hab_cmd" pkg hash "$pkg_artifact")
+
+    @"
+pkg_origin=$pkg_origin
+pkg_name=$pkg_name
+pkg_version=$pkg_version
+pkg_release=$pkg_release
+pkg_ident=${pkg_origin}/${pkg_name}/${pkg_version}/${pkg_release}
+pkg_artifact=$(Split-Path $pkg_artifact -Leaf)
+pkg_sha256sum=$_pkg_sha256sum
+pkg_blake2bsum=$_pkg_blake2bsum
+"@ | Out-File "$pkg_output_path\last_build.env" -Encoding ascii
+}
+
+# A function for cleaning up after yourself. Delegates most of the
+# implementation to the `Invoke-DefaultEnd` function.
+function Invoke-End {
+  Invoke-DefaultEnd
+}
+
+# Default implementation for the `Invoke-End` phase.
+function Invoke-DefaultEnd {
+}
+
+
+# # Main Flow
+
+# Expand the context path to an absolute path
+if (-Not (Test-Path "$Context")) {
+    _Exit-With "Context must be an existing directory" 10
+}
+$script:PLAN_CONTEXT = (Get-Item $Context).FullName
+
+# Now to ensure a `plan.ps1` exists where we expect it. There are 2 possible
+# candidate locations relative to the `$PLAN_CONTEXT` directory: a `./plan.ps1`
+# or a `./habitat/plan.ps1`. Only one or the other location is allowed so that
+# a Plan author isn't confused if they edit one to have this program read
+# the other.
+
+# We'll make sure that both are not present, and if so abort.
+if ((Test-Path "$PLAN_CONTEXT\plan.ps1") -and (Test-Path "$PLAN_CONTEXT\habitat\plan.ps1")) {
+    $places = "$PLAN_CONTEXT\plan.sh and $PLAN_CONTEXT\habitat/plan.sh"
+    _Exit-With "A Plan file was found at $places. Only one is allowed at a time" 42
+}
+# We check if the provided path has a `plan.sh` in it in either location. If
+# not, we'll quickly bail.
+if (-Not (Test-Path "$PLAN_CONTEXT\plan.ps1")) {
+    if (Test-Path "$PLAN_CONTEXT\habitat\plan.ps1") {
+        $PLAN_CONTEXT = "$PLAN_CONTEXT\habitat"
+    } else {
+        $places = "$PLAN_CONTEXT\plan.sh or $PLAN_CONTEXT\habitat/plan.sh"
+        _Exit-With "Plan file not found at $places" 42
+    }
+}
+
+# Change into the `$PLAN_CONTEXT` directory for proper resolution of relative
+# paths in `plan.ps1`
+Set-Location "$PLAN_CONTEXT"
+# @TODO fin - is there a way to change directory only in this program without
```

Comment:
powershell has `push-location` and `pop-location` both aliased by `pushd` and `popd`. You can wrap in a `try {}` and put the `pop` in a `finally {}` to achieve the above.